crypto/s2n_certificate.c

```diff
@@ -605,3 +611,108 @@ int s2n_get_cert_der(const struct s2n_cert *cert, const uint8_t **out_cert_der,
 
     return S2N_SUCCESS;
 }
+
+static int s2n_buffer_free(uint8_t** data)
+{
+    if (*data != NULL) {
+        free(*data);
+    }
+    return S2N_SUCCESS;
+}
+
+int s2n_get_utf8_string_from_extension_data(const uint8_t *extension_data, uint32_t extension_len, uint8_t **out_data, uint32_t *out_len)
+{
+    POSIX_ENSURE_REF(extension_data);
+    POSIX_ENSURE_REF(out_data);
+    POSIX_ENSURE_REF(out_len);
+
+    *out_len = extension_len - OCTET_STRING_OFFSET;
+    POSIX_ENSURE_GT(*out_len, 0);
+
+    DEFER_CLEANUP(uint8_t *temp_buf = NULL, s2n_buffer_free);
+    temp_buf = malloc(sizeof(unsigned char) * (*out_len));
```

Comment:
You should be using blobs for allocation so they go through the memory callbacks. It also means you don't have to define your own free function

Reply:
This _free is to be called by the caller after they use the returned value  and `s2n_blob` is an internal structure.

Reply:
Then we shouldn't be allocating anything at all and the caller should be passing a pre-allocated buffer in to the function.